File: src/Python/229.MajorityElementII.py

class Solution:
    def majorityElement(self, nums):
        major1 = 0
        cnt1 = 0
        major2 = 0
        cnt2 = 0
        n = len(nums)
        for i in range(n):
            if cnt1 == 0 and major2 != nums[i]:
                major1 = nums[i]
                cnt1 = 1
            elif cnt2 == 0 and major1 != nums[i]:
                major2 = nums[i]
                cnt2 = 1
            elif major1 == nums[i]:
                cnt1 += 1
            elif major2 == nums[i]:
                cnt2 += 1
            else:
                cnt1 -= 1
                cnt2 -= 1
        cnt1 = 0
        cnt2 = 0
        for i in range(n):
            if nums[i] == major1:
                cnt1 += 1
            elif nums[i] == major2:
                cnt2 += 1
        ans = []
        if cnt1 > n // 3:
            ans.append(major1)
        if cnt2 > n // 3:
            ans.append(major2)
        return ans


if __name__ == "__main__":
    solution = Solution()
    print(solution.majorityElement([3, 2, 3]))
    print(solution.majorityElement([1, 1, 1, 3, 3, 2, 2, 2]))

File: src/CSharp/229.MajorityElementII.cs

using System;
using System.Collections.Generic;

namespace PeefyLeetCode.MajorityElementII
{
public class Solution {
    public IList<int> MajorityElement(int[] nums) {
        int major1 = 0;
        int cnt1 = 0;
        int major2 = 0;
        int cnt2 = 0;
        int n = nums.Length;
        for (int i = 0; i < n; i++){
            if (cnt1 == 0 && major2 != nums[i]){
                major1 = nums[i];
                cnt1 = 1;
            }
            else if (cnt2 == 0 && major1 != nums[i]){
                major2 = nums[i];
                cnt2 = 1;
            }
            else if (major1 == nums[i])
                cnt1 += 1;
            else if (major2 == nums[i])
                cnt2 += 1;
            else{
                cnt1 -= 1;
                cnt2 -= 1;
            }
        }           
        cnt1 = 0;
        cnt2 = 0;
        for (int i = 0; i < n; i++){
            if (nums[i] == major1)
                cnt1 += 1;
            else if (nums[i] == major2)
                cnt2 += 1;
        }
        List<int> ans = new List<int>();
        if (cnt1 > n / 3)
            ans.Add(major1);
        if (cnt2 > n / 3)
            ans.Add(major2);
        return ans;
    }
}
}